### PR TITLE
Invert hide email logic to be possitive by default.

### PR DESCRIPTION
**Pull request by simi**
This is motivated by https://github.com/rubygems/rubygems.org/pull/3287 and is initial pull request in serie. For now my intention is to revert the wording and logic in app. Email should be private by default and user should opt-in to make it public if welcomed.

This is currently true in app logic, but the wording around seems opposite. Also it is not known during registration if email is going to be publicly visible. With this change it should be clear.

In next step I would like to introduce additional checkbox to opt-in into using gravatar to handle https://github.com/rubygems/rubygems.org/issues/3278 and https://github.com/rubygems/rubygems.org/pull/3287.

And in the end I would like to inline profile edit form and sign-up form into one partial providing all options and columns in both. IMHO there is big chance to get even optional fields like Twitter handle filled during registration. It seems bad UX to register with just few fields and need to complete profile at profile edit page (since there is currently no way to let new user know profile could be completed later).

In the end I would like to revisit some users columns and add `NOT NULL` constraint (for example to `public_email` one), to prevent potential 3 states on boolean column.

---

- renames hide_email boolean into public_email
- makes email hidden by default
- add option to optionally show email in public profile during registration

### new checkbox in sign up form

image

/cc @damien @jchestershopify

### CODE REVIEW

**Review by damien — COMMENTED** [review]
I'm not a core contributor and likewise cannot approve this myself, but I do think this is a great idea. Changes seem straightforward and this is mainly a semantic change that simplifies logic around public/private emails.

**Review by indirect — APPROVED** [review]
looks good to me 👍🏻

File: db/migrate/20230423202740_hide_email_by_default.rb
Comment (segiddins) [inline db/migrate/20230423202740_hide_email_by_default.rb:1]: I would prefer we do this in multiple steps:

1) add new `public_email` column
2) dual write both columns
3) backfill new column
4) read only from new column
5) delete old column

otherwise, there is the potential that any mistake will leave us with irrecoverable data changes

Comment (simi) [reply db/migrate/20230423202740_hide_email_by_default.rb:1]: If I understand it well, whole up/down blocks are wrapped in transaction and PostgreSQL schema is mostly transactional. That means whole block is executed or nothing leaving no space for unpredictable state. :thinking: But I can change to suggested format if needed.

**Review by jenshenny — APPROVED** [review]
Thanks @simi! This and the future steps makes sense to me.

File: db/migrate/20230423202740_hide_email_by_default.rb
Comment (jenshenny) [reply db/migrate/20230423202740_hide_email_by_default.rb:1]: I don't feel too strongly but I also prefer to split these kinds of data changes into multiple steps as @segiddins described.

File: app/avo/resources/user_resource.rb
Comment (bettymakes) [inline app/avo/resources/user_resource.rb:29]: Hey @simi! I was working on a similar change. @segiddins astutely pointed out that this _may_ [break the old audits](https://github.com/rubygems/rubygems.org/pull/3847#discussion_r1241257409). Have you had a chance to investigate this as well? 

Based on my preliminary investigations today, it looks like the audits would indeed get mucked up. I wanted to check in to confirm whether you came to a similar conclusion.

If so, I'll dig into the Avo docs later this week to see if they have any guidance on how they handle renaming of fields.

Comment (bettymakes) [reply app/avo/resources/user_resource.rb:29]: Hm, with further investigations, it appears the audits aren't broken after all. I'd still like to do a sanity check with you Josef to affirm this finding.

File: db/migrate/20230423202740_hide_email_by_default.rb
Comment (simi) [reply db/migrate/20230423202740_hide_email_by_default.rb:1]: @segiddins I'm about to merge this. If I understand it well (by recent experience), we will need to do it exactly as you mentioned (using `ignored_columns` trick) to make smooth deployment. Can you confirm please? I'll update the PR.

Comment (segiddins) [reply db/migrate/20230423202740_hide_email_by_default.rb:1]: please rebase & see if strong_migrations errors

Comment (simi) [reply db/migrate/20230423202740_hide_email_by_default.rb:1]: It does not by default, since this migration has old timestamp. But setting `StrongMigrations.start_after = 0` and trying to run this migration errors thanks to `strong_migrations`.

**Comment by simi** [comment]
:information_source: this needs to be merged after https://github.com/rubygems/rubygems.org/pull/4035/ (it is included in this PR as well to keep it together) is deployed and related task finishes.

**Review by martinemde — APPROVED** [review]
I can't comment on the audits but the code looks more clear. If we need to keep a hide_email method around for audits, that's not a big deal.

**Comment by simi** [comment]
> I can't comment on the audits but the code looks more clear. If we need to keep a hide_email method around for audits, that's not a big deal.

There will be additional PR to remove the column. We can check how that affects auditing feature and what we can do about that to not block unrelated changes (like this) by auditing feature.

File: app/avo/resources/user_resource.rb
Comment (simi) [reply app/avo/resources/user_resource.rb:30]: column removal will be handled in following PR (not still clear in which way), marking this as resolved

**Pull request merged**
